pkg/sql/catalog/tabledesc/constraint.go

```diff
@@ -11,215 +11,477 @@
+	if columnOrdinal < 0 || columnOrdinal >= len(c.desc.ColumnIDs) {
+		return 0
 	}
+	return c.desc.ColumnIDs[columnOrdinal]
+}
 
-	// addConstraintToSetsByValidity is a function that adds constraint `c` to various slice
-	// categorized by validity.
-	addConstraintToSetsByValidity := func(
-		c catalog.Constraint,
-		all, allActiveAndInactive, allActive []catalog.Constraint,
-	) (
-		updatedAll []catalog.Constraint,
-		updatedAllActiveAndInactive []catalog.Constraint,
-		updatedAllActive []catalog.Constraint,
-	) {
-		cstValidity := c.GetConstraintValidity()
-		all = addIfNotExists(c, all, nil)
-
-		if cstValidity == descpb.ConstraintValidity_Validated ||
-			cstValidity == descpb.ConstraintValidity_Validating ||
-			cstValidity == descpb.ConstraintValidity_Unvalidated {
-			allActiveAndInactive = addIfNotExists(c, allActiveAndInactive, nil)
-		}
-		if cstValidity == descpb.ConstraintValidity_Validated {
-			allActive = addIfNotExists(c, allActive, nil)
-		}
-		return all, allActiveAndInactive, allActive
+// CollectReferencedColumnIDs implements the catalog.CheckConstraint
+// interface.
+func (c checkConstraint) CollectReferencedColumnIDs() catalog.TableColSet {
+	return catalog.MakeTableColSet(c.desc.ColumnIDs...)
+}
+
+// IsNotNullColumnConstraint implements the catalog.CheckConstraint interface.
+func (c checkConstraint) IsNotNullColumnConstraint() bool {
+	return c.desc.IsNonNullConstraint
+}
+
+// IsHashShardingConstraint implements the catalog.CheckConstraint interface.
+func (c checkConstraint) IsHashShardingConstraint() bool {
+	return c.desc.FromHashShardedColumn
+}
+
+// GetConstraintID implements the catalog.Constraint interface.
+func (c checkConstraint) GetConstraintID() descpb.ConstraintID {
+	return c.desc.ConstraintID
+}
+
+// GetConstraintValidity implements the catalog.Constraint interface.
+func (c checkConstraint) GetConstraintValidity() descpb.ConstraintValidity {
+	return c.desc.Validity
+}
+
+// IsConstraintValidated implements the catalog.Constraint interface.
+func (c checkConstraint) IsConstraintValidated() bool {
+	return c.desc.Validity == descpb.ConstraintValidity_Validated
+}
+
+// IsConstraintUnvalidated implements the catalog.Constraint interface.
+func (c checkConstraint) IsConstraintUnvalidated() bool {
+	return c.desc.Validity == descpb.ConstraintValidity_Unvalidated
+}
+
+// GetName implements the catalog.Constraint interface.
+func (c checkConstraint) GetName() string {
+	return c.desc.Name
+}
+
+// AsCheck implements the catalog.ConstraintProvider interface.
+func (c checkConstraint) AsCheck() catalog.CheckConstraint {
+	return &c
+}
+
+// String implements the catalog.Constraint interface.
+func (c checkConstraint) String() string {
+	return fmt.Sprintf("%+v", c.desc)
+}
+
+// IsEnforced implements the catalog.Constraint interface.
+func (c checkConstraint) IsEnforced() bool {
+	return !c.IsMutation() || c.WriteAndDeleteOnly()
```

Comment:
Unfortunately, the legacy code did a bad job about the state machine of constraints while it's still on the mutation slice. I think it's better to check the constraint mutation's validity instead of the MutationDescriptor_State, so I suggest we instead just do `return c.desc.validity != dropping`. But this again iterates one of my comments above that maybe we can get rid of this method and just expose `GetConstraintValidity`.

Reply:
To illustrate what I mean above a bit more: consider adding a check constraint. When we enqueue the check constraint onto the mutation slice, its MutationDescriptor_State will be (unfortunately) DELETE_ONLY but its validity is VALIDATING, which should be considered as Enforced but this method now will incorrectly returns false.

Reply:
A constraint's validity and whether it's enforced are two different things:
- the validity concerns the data that was present _before_ the constraint is added;
- a constraint being enforced or not concerns the data that appears _after_ the constraint is added.

Nonetheless